minimalkv/_get_store.py

```diff
@@ -32,27 +41,110 @@ def get_store_from_url(url: str) -> KeyValueStore:
+
+    scheme_to_store: Dict[str, Type[KeyValueStore]] = {
+        "s3": S3FSStore,
+        "hs3": HS3FSStore,
+        "boto": HS3FSStore,
+    }
+
+    parsed_url = urisplit(url)
+    # Wrappers can be used to add functionality to a store, e.g. encryption.
+    # Wrappers are separated by `+` and can be specified in two ways:
+    # 1. As the fragment, e.g. "s3://...#wrap:readonly"
+    # 2. Deprecated: As part of the scheme, e.g. "s3+readonly://..."
+    wrappers = _extract_wrappers(parsed_url)
+
+    # Remove wrappers from scheme
+    scheme_parts = parsed_url.getscheme().split("+")
+    # pop off the type of the store
+    scheme = scheme_parts[0]
+
+    if scheme not in scheme_to_store:
+        # If we can't find the scheme, we fall back to the old creation methods
+        return get_store(**url2dict(url))
+
+    store_cls_from_url = scheme_to_store[scheme]
+    if store_cls is not None and store_cls_from_url != store_cls:
+        raise ValueError(
+            f"URL scheme {scheme} does not match store class {store_cls.__name__}"
+        )
+
+    query_listdict: Dict[str, List[str]] = parsed_url.getquerydict()
+    # We will just use the last occurrence for each key
+    query = {k: v[-1] for k, v in query_listdict.items()}
+
+    store = store_cls_from_url._from_parsed_url(parsed_url, query)
+
+    # apply wrappers/decorators:
+    from minimalkv._store_decoration import decorate_store
+
+    wrapped_store = reduce(decorate_store, wrappers, store)
+
+    return wrapped_store
+
 
-    Get the encoded credentials as string like so:
+def _extract_wrappers(parsed_url: SplitResult) -> List[str]:
+    """
+    Extract wrappers from a parsed URL.
+
+    Wrappers allow you to add additional functionality to a store, e.g. encryption.
+    Wrappers are specified in the fragment part of the URL, e.g. "s3://...#wrap:readonly+urlencode"
 
-    .. code-block:: python
+    Deprecated: wrappers can also be specified as part of the scheme, e.g. "s3+readonly+urlencode://...".
+    This way of specifying wrappers will be removed in a future version.
 
-    from pathlib import Path
-    import base64
-    json_as_bytes = Path(<path_to_json>).read_bytes()
-    json_b64_encoded = base64.urlsafe_b64encode(b).decode()
+    Parameters
+    ----------
+    parsed_url: SplitResult
+        The parsed URL.
 
+    Returns
+    -------
+    wrappers: List[str]
+        The list of wrappers.
     """
-    return get_store(**url2dict(url))
+    # split off old-style wrappers, if any:
+    parts = parsed_url.getscheme().split("+")
+    # pop off the type of the store
+    parts.pop(0)
+    old_wrappers = list(reversed(parts))
+
+    if old_wrappers:
+        warnings.warn(
+            """
+            Using wrappers as part of the scheme is deprecated. Please specify wrappers as part of the fragment instead,
+            e.g. "s3://...#wrap:readonly+urlencode" instead of "s3+readonly+urlencode://...".
+            """,
+            DeprecationWarning,
+        )
```

Comment:
This was already previously deprecated, right? It's nothing we've introduced in this PR, right?

Reply:
There was no warning given previously, so we should probably keep it like that for now.